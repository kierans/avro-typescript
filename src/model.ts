/**** Contains the Interfaces and Type Guards for Avro schema */

export type Schema = RecordType | EnumType;
export interface ConversionOptions {
    logicalTypes?: { [type: string]: string };
}

export type Type = NameOrType | NameOrType[];
export type NameOrType = TypeNames | RecordType | ArrayType | NamedType | LogicalType;
export type TypeNames = "record" | "array" | "null" | "map" | string;

export interface Field {
    name: string;
    type: Type;
    default?: string | number | null | boolean;
}

export interface BaseType {
    type: TypeNames;
}

export interface RecordType extends BaseType {
    type: "record";
    name: string;
    fields: Field[];
}

export interface ArrayType extends BaseType {
    type: "array";
    items: Type;
}

export interface MapType extends BaseType {
    type: "map";
    values: Type;
}

export interface EnumType extends BaseType {
    type: "enum";
    name: string;
    symbols: string[];
}

export interface NamedType extends BaseType {
    type: string;
}

export interface LogicalType extends BaseType {
    logicalType: string;
}

export function isRecordType(type: BaseType): type is RecordType {
    return type.type === "record";
}

export function isArrayType(type: BaseType): type is ArrayType {
    return type.type === "array";
}

export function isMapType(type: BaseType): type is MapType {
    return type.type === "map";
}

export function isEnumType(type: BaseType): type is EnumType {
    return type.type === "enum";
}

export function isUnionType(type: Type): type is NameOrType[] {
    return type instanceof Array;
}

export function isOptional(type: Type): boolean {
    if (isUnionType(type)) {
        const t1 = type[0];
        if (typeof t1 === "string") {
            return t1 === "null";
        }
    }
}

export function isPrimitiveType(type: Type): type is string {
    return typeof type === "string";
}

export function isLogicalType(type: Type): type is LogicalType {
    return typeof type !== "string" && "logicalType" in type;
}
