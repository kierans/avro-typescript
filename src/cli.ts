#! /usr/bin/env node

import WritableStream = NodeJS.WritableStream;
import * as fs from "fs";

import { Command } from 'commander';

import { Schema, } from "./model";
import { avroToTypeScript } from "./index";

interface Opts {
  input?: string;
  output?: string;
}

main();

async function main() {
  const program = new Command();
  program
    .option("-i, --input <file>", "Input file. If not present, defaults to STDIN")
    .option("-o, --output <file>", "Output file. If not present, defaults to STDOUT");

  const opts = program.parse(process.argv).opts();
  const schema = await readSchema(opts);
  const ts = avroToTypeScript(schema).split("\n");

  await writeTypes(opts, ts);
}

function readSchema(opts: Opts): Promise<Schema> {
  const stream = opts.input ? fs.createReadStream(opts.input) : process.stdin;
  stream.setEncoding("utf-8");

  return new Promise((resolve, reject) => {
    const chunks = [];

    stream.on("error", reject);

    stream.on("data", (chunk) => {
      chunks.push(chunk);
    });

    stream.on("end", () => {
      resolve(JSON.parse(chunks.join("")))
    });
  });
}

function writeTypes(opts: Opts, data: string[]): Promise<void> {
  const stream = opts.output ? fs.createWriteStream(opts.output, { encoding: "utf-8" }) : process.stdout;

  return writeToStream(stream, data);
}

function writeToStream(stream: WritableStream, data: string[]): Promise<void> {
  return new Promise((resolve, reject) => {
    let isFull = false;

    stream.once("error", (err) => {
      // we don't care about draining any more.
      stream.removeAllListeners("drain");

      reject(err);
    });

    for (let lineNumber = 0; lineNumber < data.length; lineNumber++) {
      const line = data[lineNumber];
      isFull = !(stream.write(`${line}\n`));

      if (isFull) {
        stream.once("drain", () => {
          // remove the previously assigned error handler as we don't need it now
          stream.removeAllListeners("error");

          // add 1 because we've written the current line
          writeToStream(stream, data.slice(lineNumber + 1))
            .then(resolve, reject);
        });

        return;
      }
    }

    resolve();
  });
}
