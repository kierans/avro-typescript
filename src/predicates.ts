export const isEqual: (a: any) => (b: any) => boolean =
  (a) => (b) => a === b;
