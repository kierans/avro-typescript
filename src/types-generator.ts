import {
  ConversionOptions,
  EnumType,
  Field,
  Metadata,
  NameOrType,
  PrimitiveTypeNames,
  RecordType,
  Type,
  fullName,
  isArrayType,
  isEnumType,
  isLogicalType,
  isMapType,
  isNamedType,
  isPrimitiveType,
  isRecordType,
  isReferencedType,
  isUnionType,
} from "./model";

/** Convert a primitive type from avro to TypeScript */
export function convertPrimitive(_: Metadata, avroType: PrimitiveTypeNames): string {
  switch (avroType) {
    case "long":
    case "int":
    case "double":
    case "float":
      return "number";

    case "bytes":
      return "Buffer";

    case "null":
      return "null";

    case "boolean":
      return "boolean";

    case "string":
      return "string";

    default:
      return "UNKNOWN_PRIMITIVE";
  }
}

/** Convert an Avro Record type. Return the name, but add the definition to the file */
export function convertRecord(
  meta: Metadata,
  recordType: RecordType,
  fileBuffer: string[],
  opts: ConversionOptions
): string {
  let buffer = `export interface ${recordType.name} {\n`;
  for (let field of recordType.fields) {
    buffer += convertFieldDec(meta, field, fileBuffer, opts) + "\n";
  }
  buffer += "}\n";
  fileBuffer.push(buffer);
  return recordType.name;
}

/** Convert an Avro Enum type. Return the name, but add the definition to the file */
export function convertEnum(_: Metadata, enumType: EnumType, fileBuffer: string[]): string {
  const enumDef = `export enum ${enumType.name} { ${enumType.symbols.map(sym => `${sym} = '${sym}'`).join(", ")} };\n`;
  fileBuffer.push(enumDef);
  return enumType.name;
}

export function convertUnionType(meta: Metadata, type: NameOrType): NameOrType  {
  if (isReferencedType(type)) {
    return discriminatorType(meta, type, type);
  }

  if (isNamedType(type)) {
    return discriminatorType(meta, type.name, type);
  }

  return type;
}

export function convertType(meta: Metadata, type: Type, buffer: string[], opts: ConversionOptions): string {
  // if it's just a name, then use that
  if (isReferencedType(type)) {
    return type;
  }

  if (isPrimitiveType(type)) {
    return convertPrimitive(meta, type);
  }

  if (isUnionType(type)) {
    const discriminatedTypes = type.map((t) => convertUnionType(meta, t));

    // array means a Union. Use the names and call recursively
    return discriminatedTypes.map((t) => convertType(meta, t, buffer, opts)).join(" | ");
  }

  if (isRecordType(type)) {
    // record, use the name and add to the buffer
    return convertRecord(meta, type, buffer, opts);
  }

  if (isArrayType(type)) {
    // array, call recursively for the array element type
    return convertType(meta, type.items, buffer, opts) + "[]";
  }

  if (isMapType(type)) {
    // Dictionary of types, string as key
    return `{ [index:string]:${convertType(meta, type.values, buffer, opts)} }`;
  }

  if (isEnumType(type)) {
    // array, call recursively for the array element type
    return convertEnum(meta, type, buffer);
  }

  if (isLogicalType(type)) {
    if (opts.logicalTypes && opts.logicalTypes[type.logicalType]) {
      return opts.logicalTypes[type.logicalType];
    }

    return convertType(meta, type.type, buffer, opts);
  }

  console.error("Cannot work out type", type);

  return "UNKNOWN";
}

export function convertFieldDec(meta: Metadata, field: Field, buffer: string[], opts: ConversionOptions): string {
  // Union Type
  return `\t${field.name}: ${convertType(meta, field.type, buffer, opts)};`;
}

export function discriminatorType(meta: Metadata, name: string, type: NameOrType): RecordType {
  return {
    type: "record",
    name: `${name}Discriminator`,
    fields: [
      {
        name: `"${fullName(meta.namespace, name)}"`,
        type
      }
    ]
  }
}
