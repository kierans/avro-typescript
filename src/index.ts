import { convertEnum, convertRecord } from "./types-generator";

export {
    EnumType,
    Field,
    isArrayType,
    isEnumType,
    isLogicalType,
    isMapType,
    isOptional,
    isRecordType,
    RecordType,
    Type,
} from "./model";

import {
    ComplexType,
    ConversionOptions,
    isEnumType,
    isNamedType,
    isRecordType,
    Metadata,
    Schema,
} from "./model";

/** Converts an Avro record type to a TypeScript file */
export function avroToTypeScript(schema: Schema, opts: ConversionOptions = {}): string {
    const metadata: Metadata = !isNamedType(schema) ? {} : {
        namespace: schema.namespace
    };

    return convertSchema(metadata, schema, opts).join("\n");
}

function convertSchema(meta: Metadata, schema: Schema, opts: ConversionOptions): string[] {
    const output: string[] = [];

    if (isEnumType(schema)) {
        convertEnum(meta, schema, output);

        return output;
    }

    if (isRecordType(schema)) {
        convertRecord(meta, schema, output, opts);

        return output;
    }

    throw "Unknown top level type " + (schema as ComplexType)["type"];
}
