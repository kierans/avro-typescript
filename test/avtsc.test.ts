import * as fs from "fs";
import { avroToTypeScript } from "../src";
import { RecordType, Schema } from "../src/model";

describe("avroToTypeScript", () => {
    test("it should generate an interface", () => {
        const schemaText = fs.readFileSync(__dirname + "/example.avsc", "utf-8");
        const schema = JSON.parse(schemaText);
        expect(avroToTypeScript(schema as RecordType)).toMatchSnapshot();
    });

    test("it should correctly type strings with logicalType", () => {
        const schema: Schema = {
            type: "record",
            name: "datedRecord",
            fields: [
                {
                    name: "eventDate",
                    type: {
                        type: "string",
                        logicalType: "iso-datetime",
                    },
                },
            ],
        };
        expect(avroToTypeScript(schema)).not.toEqual(expect.stringContaining("UNKNOWN"));
    });

    test("it should support overriding logical types", () => {
        const schema: Schema = {
            type: "record",
            name: "logicalOverrides",
            fields: [
                {
                    name: "eventDate",
                    type: {
                        type: "int",
                        logicalType: "date",
                    },
                },
                {
                    name: "startTime",
                    type: {
                        type: "int",
                        logicalType: "timestamp-millis",
                    },
                },
                {
                    name: "displayTime",
                    type: {
                        type: "string",
                        logicalType: "iso-datetime",
                    },
                },
            ],
        };
        const actual = avroToTypeScript(schema, {
            logicalTypes: {
                date: 'Date',
                'timestamp-millis': 'Date',
            }
        });
        expect(actual).toMatchSnapshot();
        expect(actual).not.toEqual(expect.stringContaining("eventDate: number"));
        expect(actual).not.toEqual(expect.stringContaining("startTime: number"));
        expect(actual).toEqual(expect.stringContaining("displayTime: string"));
    });
});

describe("enumToTypesScript", () => {
    test("it should generate an enum", () => {
        const schemaText = fs.readFileSync(__dirname + "/just-enum.avsc", "utf-8");
        const schema = JSON.parse(schemaText);
        expect(avroToTypeScript(schema as RecordType)).toMatchSnapshot();
    });

    test("it should correctly type strings with logicalType", () => {
        const schema: Schema = {
            type: "enum",
            name: "CompanySize",
            symbols: ["SMALL", "MEDIUM", "LARGE"],
        };

        expect(avroToTypeScript(schema)).not.toEqual(expect.stringContaining("UNKNOWN"));
    });
});
