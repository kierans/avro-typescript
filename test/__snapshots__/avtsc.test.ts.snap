// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`avroToTypeScript it should generate an interface 1`] = `
"export interface StringDiscriminator {
	string: string;
}

export enum numberEnumType { ONE = 'ONE', TWO = 'TWO' };

export interface NumberEnumTypeDiscriminator {
	\\"numberEnumType\\": numberEnumType;
}

export interface DoubleDiscriminator {
	double: number;
}

export interface NumberEnumTypeMapDiscriminator {
	map: { [index:string]:numberEnumType };
}

export interface LongDiscriminator {
	long: number;
}

export interface AllocsType {
	allocAccount: null | StringDiscriminator;
	noNestedPartyIDs: null | LongDiscriminator;
	allocQty: null | DoubleDiscriminator;
}

export interface AllocsTypeArrayDiscriminator {
	array: AllocsType[];
}

export interface AllocsTypeDiscriminator {
	\\"AllocsType\\": AllocsType;
}

export interface ExampleType {
	unionEnum: null | StringDiscriminator | NumberEnumTypeDiscriminator;
	mandatoryString: string;
	adouble: null | DoubleDiscriminator;
	astring: null | StringDiscriminator;
	amap: null | StringDiscriminator | NumberEnumTypeMapDiscriminator;
	recordArray: null | AllocsTypeArrayDiscriminator;
	processCode: null | StringDiscriminator;
	named: null | AllocsTypeDiscriminator;
}
"
`;

exports[`avroToTypeScript it should support overriding logical types 1`] = `
"export interface logicalOverrides {
	eventDate: Date;
	startTime: Date;
	displayTime: string;
}
"
`;

exports[`enumToTypesScript it should generate an enum 1`] = `
"export enum CompanySize { SMALL = 'SMALL', MEDIUM = 'MEDIUM', LARGE = 'LARGE' };
"
`;
